book/numerical_methods/curve_fitting.nim

```diff
@@ -29,6 +31,109 @@ yes.
+[Curve fitting](https://en.wikipedia.org/wiki/Curve_fitting) is the task of finding the "best" parameters for a given function,
+such that it minimizes the error on the given data. The simplest example being,
+finding the slope and intersection of a line $y = ax + b$ using such that it minimizes the squared errors
+between the data points and the line.
+[numericalnim](https://github.com/SciNim/numericalnim) implements the [Levenberg-Marquardt algorithm](https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm)(`levmarq`)
+for solving non-linear least squares problems. One such problem is curve fitting, which aims to find the parameters that (locally) minimizes
+the error between the data and the fitted curve.
+
+The required imports are:
+"""
+
+nbCode:
+  import numericalnim, ggplotnim, arraymancer, std / [math, sequtils]
+
+nbText: hlMd"""
+In some cases you know the actual form of the function you want to fit,
+but in other cases you may have to guess and try multiple different ones.
+In this tutorial we will assume we know the form but it works the same regardless.
+The test curve we will sample points from is
+$$f(t) = \alpha + sin(\beta t + \gamma) e^{-\delta t}$$
+with $\alpha = 0.5, \beta = 6, \gamma = 0.1, \delta = 1$. This will be a decaying sinus wave with an offset.
+We will add a bit a noise to it as well:
+"""
+
+nbCode:
+  proc f(alpha, beta, gamma, delta, t: float): float =
+    alpha + sin(beta * t + gamma) * exp(-delta*t)
+
+  let
+    alpha = 0.5
+    beta = 6.0
+    gamma = 0.1
+    delta = 1.0
+  let t = arraymancer.linspace(0.0, 3.0, 20)
+  let yClean = t.map_inline:
+    f(alpha, beta, gamma, delta, x)
+  let y = yClean + randomTensor(t.shape[0], 0.05)
+
+let tDense = arraymancer.linspace(0.0, 3.0, 200)
+let yDense = tDense.map_inline:
+  f(alpha, beta, gamma, delta, x)
+
+block:
+  let df = toDf(t, y, tDense, yDense)
+  df.ggplot(aes("t", "y")) +
+    geom_point() +
+    geom_line(aes("tDense", "yDense")) +
+    ggsave("images/levmarq_rawdata.png")
+
+nbImage("images/levmarq_rawdata.png")
+
+nbText: hlMd"""
+Here we have the original function along with the sampled points with noise.
+Now we have to create a proc that `levmarq` expects. Specifically does it
+want all the parameters in a `Tensor` instead of by themselves: 
+"""
+
+nbCode:
+  proc fitFunc(params: Tensor[float], t: float): float =
+    let alpha = params[0]
+    let beta = params[1]
+    let gamma = params[2]
+    let delta = params[3]
+    result = f(alpha, beta, gamma, delta, t)
+
+nbText: hlMd"""
+As we can see, all the parameters that we want to fit, are passed in as
+a single 1D Tensor that we unpack for clarity here. The only other thing
+that is needed is an initial guess of the parameters. We will set them to all 1 here: 
+"""
+
+nbCode:
+  let initialGuess = ones[float](4)
+
+nbText: "Now we are ready to do the actual fitting:"
+
+nbCode:
+  let solution = levmarq(fitFunc, initialGuess, t, y)
+  echo solution
```

Comment:
The simple answer is that I don't know how to do that :P (Or rather haven't researched how to do it :sweat_smile:). If you have any good resources I could try and add it :+1:

Reply:
I don't have any good resources from the top of my head. But I guess any good statistics book should cover it (just checked and Barlow Statistics certainly does; I guess Statistical Data Analysis by Cowan will too). 

But generally at least χ² is trivial (you can just compute it at the end using the data points and fitted points as `χ² = Σ_i (y_i - ŷ_i) / (Δy_i)` and the degrees of freedom (dof) are simply `# of data points - # of fit parameters`. 

Allowing for y uncertainties in the fit (it depends a bit on your implementation of Levenberg Marquardt (haven't looked at it and I've never implemented it after all), but generally speaking _if_ your optimization was a direct optimization of the χ² as computed like above you would simply take y uncertainties into account by plugging them in there. If none given you'd just give them a weight of 1.

How to compute the covariance matrix I'd need to look up myself, oops.

Reply:
Thanks, I'll read up a bit on that then :+1: 

The formulas I find online differ from your though: `χ² = Σ (y_i - ŷ_i)² / ŷ_i`. And this looks very much like a relative error of sorts. `levmarq` minimized the mean squared error directly, so it is the absolute error instead. So this might take a while to untangle :sweat_smile:

Reply:
Never mind my last comment, found the formula with the uncertainty now as well 👍 it should be easy to incorporate the uncertainties as weights when calculating the error so that we actually minimize the Chi-squared error directly.

Reply:
Yeah, I forgot the `²` in that formula, sorry oops!

Reply:
I have added a section on the χ² now. But it is mostly based on Wikipedia, so would be nice with a pair of eyes on it that actually has a clue about it :rofl: 

As for the parameter uncertainties, I found [this link](https://stats.stackexchange.com/questions/285023/compute-standard-errors-of-nonlinear-regression-parameters-with-maximum-likeliho) which says that the covariance matrix is basically the inverse of the Hessian of the objective function. And that we can compute numerically, so I'll have a look at that tomorrow.

Reply:
I have added the parameter uncertainty now, but it feels like they are like an order of magnitude too small (I know they are just the 1-sigma error, but still). Will have to compare it to mpfit to see if the values are reasonable.

Reply:
I forgot to take the `sqrt` of the covariance so no wonder it was too small :rofl: 

Now I'm getting more reasonable results, but not the same still:

```
# Numericalnim
α = 0.4968874077122076 ± 0.001646827615777153
β = 6.037233344885453 ± 0.01065090854444737
γ = 0.09890186779685979 ± 0.006283841322723982
δ = 1.00835694586722 ± 0.008485927391057163
```

```
# MpFit
P[0] = 0.496887  +/- 0.00566763
P[1] = 6.03723   +/- 0.0367068
P[2] = 0.0989019 +/- 0.0216602
P[3] = 1.00836   +/- 0.0291869
```

I'm wondering if they are calculating it in another way. We are at least much more optimistic about the uncertainty :sunglasses:

Reply:
To finish this one up, I have implemented the parameter estimation in numericalnim now and uses them in the tutorial instead.